lib/exec.ts

```diff
@@ -316,28 +317,75 @@ export function getExecuteCEWrapperOptions(command: string, args: string[], opti
     return getCeWrapperOptions('execute', command, args, options);
 }
 
-function sandboxNsjail(command, args, options) {
+async function sanitizeDir(dir: string, root: string) {
+    logger.debug('sanitizing', {dir});
+    const dirEnts = await promises.readdir(dir, {withFileTypes: true});
+    await Promise.all(
+        dirEnts.map(async dirEnt => {
+            logger.debug('looking at:', {dirEnt});
+            const resolvedName = path.resolve(dir, dirEnt.name);
+            if (dirEnt.isSymbolicLink()) {
+                try {
+                    const dest = await fs.realpath(resolvedName);
+                    const relative = path.relative(root, dest);
+                    logger.warn(`found a symlink ${resolvedName}->${dest}; ${relative} - erasing`, {resolvedName});
+                } catch (e) {
+                    logger.warn('found a symlink but caught an exception resolving it - erasing', {
+                        dirEnt,
+                        resolvedName,
+                    });
+                }
+                await fs.unlink(resolvedName);
+            }
+            if (dirEnt.isDirectory()) await sanitizeDir(resolvedName, root);
+        }),
+    );
+}
+
+export async function sanitize(options: ExecutionOptions) {
+    // Ensure there's no symlinks that resolves outside of the working dir.
+    if (options.customCwd) {
+        const root = path.resolve(options.customCwd);
```

Comment:
Recently I discovered that dex2oat's customCwd was setup incorrectly, I fixed it ofcourse, but if we're not careful enough with PR's for new compilers/languages, it might be that customCwd is not the tmpDir but one of the binary directories where symlinks are fine. (eventhough it's setup incorrectly) - I think we could use a tiny guard here to check if customCwd is at /opt/compiler-explorer, if it is we can give a warning and not sanitize

Reply:
Great idea; I did wonder about this. `/opt/compiler-explorer` et al will be read-only _in theory_ but you're right; I should assert that this is in a "temp directory" somehow. My design isn't helpful here

Reply:
Also: if `options.appHome` is set, that should be used instead. This is in cases of for example cmake where `customCwd` is the build directory, but `options.appHome` is the actual tmpdir. (unittest example how its used https://github.com/compiler-explorer/compiler-explorer/blob/85b897e8734b9d1691c203591ddf69d2d94023a9/test/exec-tests.ts#L352)

Reply:
driveby - but we should refactor the way `appHome` and `customCwd` work.

From reading, it seems the intent was "map `/app` to `appHome` and then set CWD inside to `/app/$subdir`.

There should be a single option which controls the `--bindmount` option and a second (optional) subdirectory of `/app` to set as CWD inside the jail.